### ISSUE TITLE
TimeoutManager can postpone timeouts until endpoint restarts

### ISSUE THREAD

**Issue by tmasternak** [Particular/NServiceBus#3500]
Due to the fact that `startSlice` variable in `ExpiredTimeoutsPoller` only increases it's vulnerable to race conditions. Here is a scenario which shows how a message can get stuck in TimeoutManager until the endpoint instance is restarted:
    * It's 12:00:00 and in the `Timeouts` queue we have messageA - with dueTime=12:00:01 and message B with dueTime=12:00:02
    * MessageA can't be stored due to transient error and goes FLR
    * MessageB is stored, processed by Poller and dispatched
    * StartSlice is moved to 12:00:02 and messageA get's delivered only after endpoint restart

This might affect v5 as well (we need to verify this).

**Comment by SzymonPobiega** [comment]
Another example is when a inserting a row into the timeout storage takes a bit longer to commit. If in that time a query is executed, the next time to run the query will be retrieved based on the next timeout already in store. Then our timeout is committed but it will not be picked up until the endpoint restarts.

**Comment by MarcinHoppe** [comment]
@tmasternak @SzymonPobiega Interestingly, all persisters seem to have the `> startSlice` condition. Do you know what purpose does it serve? Is it because we might poll for timeouts while some already dispatched timeouts have not yet been removed from the store?

**Comment by SzymonPobiega** [comment]
@MarcinHoppe yes. But RabbitMQ used to have a feature that sometimes caused to omit the `startSlice` in the query to ensure nothing has been left behind.

**Comment by MarcinHoppe** [comment]
@SzymonPobiega Might this lead to duplicated messages?

**Comment by SzymonPobiega** [comment]
@MarcinHoppe no if transport and persistence can share transaction because the dispatcher removes the timeout entry in the same transaction as it dispatches the actual message. If the dispatch message comes when the corresponding timeout entry has already been removed, it is ignored.

**Comment by tmasternak** [comment]
@MarcinHoppe just to add to that. I'm might lead to duplicate control messages sent to timeout dispatcher but in the conditions mentioned by @SzymonPobiega there are no duplicates of business messages.

**Comment by MarcinHoppe** [comment]
@SzymonPobiega @tmasternak Ah, OK. Transactions to the rescue :).

**Comment by SzymonPobiega** [comment]
Here's a approach I used when building an EventStore timeout storage but not sure yet how to implement in in SQL server

 * There is a concept of *epoch*. Each timeout fits in exactly one *epoch*.
 * Epochs are identified by sequential numbers: 1, 2, 3, 4
 * Epochs have fixed length so any give time can be deterministically and without any shared context translated to an epoch number
 * Timeout sender includes in epoch. If a given epoch is marked as close, the operation is rejected and it has to reissue it with the correct (open) epoch.
 * Timeout poller queries timeouts by whole epochs in three stages:
  1. it makes the current epoch as closed
  1. it retrieves all the timeouts belonging to just closed epoch
  1. it sends all the `dispatch` messages to the dispatcher
  1. it marks that epoch as dispatched and moves to the next epoch

This was we have one bulk query per epoch and no need to look back, ever. The problem is how to model these epochs in SQL Server. One thing that comes to my mind is include an `UPDATE` in the timeout insert transaction. That `UPDATE` would have a WHERE that ensures the epoch used in insert is still open.

**Comment by tmasternak** [comment]
Another example is when processing fails here `https://github.com/Particular/NServiceBus/blob/develop/src/NServiceBus.Core/DelayedDelivery/TimeoutManager/ExpiredTimeoutsPoller.cs#L111` the timeout for which the message has not been dispatched will be skipped.

**Comment by tmasternak** [comment]
@SzymonPobiega So it looks that there are some strong transactional contraints that we need to meet e.g. transaction spanning persistence and transport. Maybe we can take advantage of that and combine epoch creation with sending messages? E.g. poller would pick up all due timeouts and assign some epoch id and dispatch message with epoch id that needs to be dispatched.

**Comment by weralabaj** [comment]
Btw did we mention that we think it might also occur in v5 and earlier versions? Should be added to description to check it.

**Comment by timbussmann** [comment]
@tmasternak did you also take https://github.com/Particular/NServiceBus/blob/develop/src/NServiceBus.Core/DelayedDelivery/TimeoutManager/StoreTimeoutBehavior.cs#L64 into account?

**Comment by SzymonPobiega** [comment]
@weralabaj I've verified that behavior in V5. Not sure about previous versions but V6 surely has the same issues.

**Comment by tmasternak** [comment]
This has not been fixed in `Core` but has been fixed in both persisters. See:
 * [RavenDB](https://github.com/Particular/NServiceBus.RavenDB/blob/master/src/NServiceBus.RavenDB/Timeouts/TimeoutPersister.cs#L37)
 * [NHibernate](https://github.com/Particular/NServiceBus.NHibernate/blob/master/src/NServiceBus.NHibernate/TimeoutPersisters/TimeoutPersister.cs#L38)
